SIP-20: WebSockets rewrite (#176)
MetaMask/SIPs: SIPS/sip-20.md

## Motivation
Currently, Snaps can only communicate with external services via HTTP requests. This limits their ability to receive real-time data updates, which is essential for many use cases, such as price feeds or event notifications. By exposing a WebSocket protocol, Snaps can establish persistent connections with external services and receive real-time updates.

## Specification

> Formal specifications are written in TypeScript. 

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

### Top-level architecture

This architecture defers WebSocket management to the client, allowing Snaps to open, close, and send messages over WebSocket. The client will handle the underlying WebSocket connections and notify the Snap of any events that occur on those connections.

> The client has multiple responsibilities in this architecture:

Reviewer: ```suggestion
The client has multiple responsibilities in this architecture:

```

- All WebSocket connections opened by Snaps MUST be monitored by the client. Each connection is identified by a unique identifier, which is returned to the Snap when the connection is opened.

- Any incoming message from WebSocket connections MUST be handled by the client and forwarded to the appropriate Snap through `onWebSocketEvent`. 

- When the Snap requests to open a WebSocket connection via `snap_openWebSocket`, the client MUST ensure that the connection is established and return a unique identifier for that connection.

- Any open connection MUST be kept alive by the client until the Snap explicitly closes it or the connection is lost, even when the Snap is terminated.

- When the Snap requests to close a WebSocket connection via `snap_closeWebSocket`, the client MUST close the connection and remove it from its list of open connections.

> - Any errors that occur during the WebSocket connection lifecycle (e.g., connection failure, message send failure) MUST be handled by the client and reported to the Snap via `onWebSocketEvent`.

Author: I'm wondering if this applies to everything. Should we for instance not throw on `snap_openWebSocket` if the initial connection fails? 🤔

```typescript

type OpenWebSocketParams = {
  url: string;
};
```
The RPC method takes one parameter:

- `url` - The URL of the WebSocket service to connect to. 
  - The URL MUST be a valid WebSocket URL, starting with `wss://`. URLs starting with `ws://` are not allowed.
  - Only one WebSocket connection can be opened per URL at a time. If a Snap tries to open a new connection to the same URL while an existing connection is still open, the client MUST throw an error.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_openWebSocket",
  params: {
    url: "wss://example.com/websocket",
  },
});

```

#### `snap_closeWebSocket`
This method allows a Snap to close an existing WebSocket connection. The method takes the unique identifier of the connection as a parameter.

```typescript
type CloseWebSocketParams = {
  id: string;
};
```
The RPC method takes one parameter:
- `id` - The unique identifier of the WebSocket connection to close. This identifier is returned by the `snap_openWebSocket` method.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_closeWebSocket",
  params: {
    id: "unique-connection-id",
  },
});
```

> #### `snap_sendMessage`

Reviewer: I think this name is too generic, especially given that everything else is specific to WebSockets.

This method allows a Snap to send a message over an existing WebSocket connection. The method takes the unique identifier of the connection and the message to send as parameters.

```typescript
type SendWebSocketMessageParams = {
  id: string;
  message: string | number[];
};
```

The RPC method takes two parameters:
- `id` - The unique identifier of the WebSocket connection to send the message over. This identifier is returned by the `snap_openWebSocket` method.

> - `message` - The message to send over the WebSocket connection. It can be either a string or a `Uint8Array`.

Reviewer: This doesn't match the type above.

```typescript
import { OnWebSocketEventHandler } from "@metamask/snap-sdk";

export const onWebSocketEvent: OnWebSocketEventHandler = async ({ event }) => {
  switch (event.type) {
    case "message":
      // Handle incoming message
      console.log(`Message received from ${event.origin}:`, event.data);
      break;
    case "open":
      // Handle connection opened
      console.log(`WebSocket connection opened with ID ${event.id} from ${event.origin}`);
      break;
    case "close":
      // Handle connection closed
      console.log(`WebSocket connection closed with ID ${event.id} from ${event.origin}`);
      break;
    case "error":
      // Handle error
      console.error(`WebSocket error occurred with ID ${event.id} from ${event.origin}`);
      break;
  }
};
```
the type for an `onWebSocketEvent` handler function's arguments is:

```typescript
export type WebSocketMessageEvent = {
  type: "message";
  id: string;
  origin: string;
  dataType: "text" | "binary";

>   data: string | Uint8Array;

Reviewer: Ideally we use a discriminated union here to determine the type of `data`, like in the previous specification.

};

export type WebSocketOpenEvent = {
  type: "open";
  id: string;
  origin: string;
};

export type WebSocketCloseEvent = {
  type: "close";
  id: string;
  origin: string;
};

export type WebSocketErrorEvent = {
  type: "error";
  id: string;
  origin: string;
};

export type WebSocketEvent =
  | WebSocketDataEvent
  | WebSocketOpenEvent
  | WebSocketCloseEvent
  | WebSocketErrorEvent;
```

```typescript
type OnWebSocketEventArgs = {
  event: WebSocketEvent;
};
```
`type` - The type of the WebSocket event, which can be one of the following:
  - `message` - Indicates that a message has been received.
  - `open` - Indicates that a WebSocket connection has been opened.
  - `close` - Indicates that a WebSocket connection has been closed.
  - `error` - Indicates that an error has occurred with the WebSocket connection.

`id` - The unique identifier of the WebSocket connection associated with the event.

`origin` - The origin of the WebSocket event.

`dataType` - The type of data received in the event, which can be either `text` or `binary`. This property is only present for `message` events.

> `data` - The data received in the event. For `message` events, this can be either a string (for text messages) or a `Uint8Array` (for binary messages). For other event types, this property is not present.

Reviewer: These shouldn't be documented here if they aren't present. Seems more clear to describe the different events separately.

> This handler does not return any value. It is used to handle WebSocket events asynchronously:

Reviewer: ```suggestion
This handler does not return any value.
```

  id: string;
};
```
The RPC method takes one parameter:
- `id` - The unique identifier of the WebSocket connection to close. This identifier is returned by the `snap_openWebSocket` method.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_closeWebSocket",
  params: {
    id: "unique-connection-id",
  },
});
```
#### `snap_sendWebSocketMessage`
This method allows a Snap to send a message over an existing WebSocket connection. The method takes the unique identifier of the connection and the message to send as parameters.

```typescript
type SendWebSocketMessageParams = {
  id: string;
  message: string | number[];
};
```

The RPC method takes two parameters:
- `id` - The unique identifier of the WebSocket connection to send the message over. This identifier is returned by the `snap_openWebSocket` method.
- `message` - The message to send over the WebSocket connection. It can be either a string or an array of bytes.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_sendWebSocketMessage",
  params: {
    id: "unique-connection-id",
    message: "Hello, WebSocket!",
  },
});
```
#### `snap_getWebSockets`

> This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier and URL of the connection.

Reviewer: The description is outdated now.

> This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier and URL of the connection.

Author: ```suggestion
This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier, the optional protocols and the URL of the connection.
```

  params: {
    id: "unique-connection-id",
  },
});
```
#### `snap_sendWebSocketMessage`
This method allows a Snap to send a message over an existing WebSocket connection. The method takes the unique identifier of the connection and the message to send as parameters.

```typescript
type SendWebSocketMessageParams = {
  id: string;
  message: string | number[];
};
```

The RPC method takes two parameters:
- `id` - The unique identifier of the WebSocket connection to send the message over. This identifier is returned by the `snap_openWebSocket` method.
- `message` - The message to send over the WebSocket connection. It can be either a string or an array of bytes.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_sendWebSocketMessage",
  params: {
    id: "unique-connection-id",
    message: "Hello, WebSocket!",
  },
});
```
#### `snap_getWebSockets`
This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier, the optional protocols and the URL of the connection.

- `id` - The unique identifier of the WebSocket connection.

> - `url` - The URL of the WebSocket connection.

Reviewer: ```suggestion
- `url` - The URL of the WebSocket connection.
- `protocols` - The optional protocols of the WebSocket connection.
```

---
sip: 20
title: WebSockets
status: Draft
author: David Drazic (@david0xd), Frederik Bolding (@frederikbolding), Guillaume Roux (@guillaumerx) 
created: 2023-12-15
updated: 2024-06-05
---

## Abstract
This SIP proposes to expose a new communication protocol to `endowment:network-access`, that enables Snaps to communicate with external services via WebSockets. This will allow Snaps to receive real-time data updates from external sources, such as price feeds or event notifications.

## Motivation

> Currently, Snaps can only communicate with external services via HTTP requests. This limits their ability to receive real-time data updates, which is essential for many use cases, such as price feeds or event notifications. By exposing a WebSocket protocol, Snaps can establish persistent connections with external services and receive real-time updates.

Author: ```suggestion
Currently, Snaps can only communicate with external services via HTTP requests. This limits their ability to receive real-time data updates, which is essential for many use cases, such as price feeds or event notifications. By exposing the WebSocket protocol, Snaps can establish persistent connections with external services and receive real-time updates.
```

  method: "snap_openWebSocket",
  params: {
    url: "wss://example.com/websocket",
    protocols: ["soap", "wamp"],
  },
});

```

#### `snap_closeWebSocket`
This method allows a Snap to close an existing WebSocket connection. The method takes the unique identifier of the connection as a parameter.

```typescript
type CloseWebSocketParams = {
  id: string;
};
```
The RPC method takes one parameter:
- `id` - The unique identifier of the WebSocket connection to close. This identifier is returned by the `snap_openWebSocket` method.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_closeWebSocket",
  params: {
    id: "unique-connection-id",
  },
});
```
#### `snap_sendWebSocketMessage`
This method allows a Snap to send a message over an existing WebSocket connection. The method takes the unique identifier of the connection and the message to send as parameters.

```typescript
type SendWebSocketMessageParams = {
  id: string;
  message: string | number[];
};
```

The RPC method takes two parameters:
- `id` - The unique identifier of the WebSocket connection to send the message over. This identifier is returned by the `snap_openWebSocket` method.

> - `message` - The message to send over the WebSocket connection. It can be either a string or a number array.

Author: ```suggestion
- `message` - The message to send over the WebSocket connection. It can be either a string or an array of bytes.
```
Perhaps?

An example of usage is given below.

```typescript
snap.request({
  method: "snap_sendWebSocketMessage",
  params: {
    id: "unique-connection-id",
    message: "Hello, WebSocket!",
  },
});
```
#### `snap_getWebSockets`
This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier and URL of the connection.

- `id` - The unique identifier of the WebSocket connection.
- `origin` - The origin of the WebSocket connection.

```typescript
type WebSocketConnection = {
  id: string;

>   origin: string;

Author: Should this be URL?

      break;
    case "open":
      // Handle connection opened
      console.log(`WebSocket connection opened with ID ${event.id} from ${event.origin}`);
      break;
    case "close":
      // Handle connection closed
      console.log(`WebSocket connection closed with ID ${event.id} from ${event.origin}`);
      break;
    case "error":
      // Handle error
      console.error(`WebSocket error occurred with ID ${event.id} from ${event.origin}`);
      break;
  }
};
```

```typescript
type OnWebSocketEventArgs = {
  event: WebSocketEvent;
};
```

```typescript
type OnWebSocketEventResponse = void;
```

#### Event Types

the type for an `onWebSocketEvent` handler function's arguments. 

```typescript
export type WebSocketEvent =
  | WebSocketMessage
  | WebSocketOpenEvent
  | WebSocketCloseEvent
  | WebSocketErrorEvent;
```

##### Message Event

This event is triggered when a message is received over a WebSocket connection. The message can be either text or binary data.

```typescript
export type WebSocketTextMessage = {
  type: "text";
  message: string;
};
export type WebSocketBinaryMessage = {
  type: "binary";

>   message: number[];

Reviewer: Wondering if this should be a base64 string, since we use that in other places for binary data.

Reviewer 2: @FrederikBolding

Author: What would be the reasoning? When I proposed this I was thinking we wanted to mirror the browser WebSocket API as much as possible

Reviewer: Just consistency with other APIs. Even better if we could have a translation layer on the execution side turning it into `Uint8Array` before passing to the Snap, but 🤷

---
sip: 20
title: WebSockets
status: Draft
author: David Drazic (@david0xd), Frederik Bolding (@frederikbolding), Guillaume Roux (@guillaumerx) 
created: 2023-12-15

> updated: 2024-06-05

Reviewer 3: ```suggestion
updated: 2025-06-05
```

- Any incoming message from WebSocket connections MUST be handled by the client and forwarded to the appropriate Snap through `onWebSocketEvent`. 

- When the Snap requests to open a WebSocket connection via `snap_openWebSocket`, the client MUST ensure that the connection is established and return a unique identifier for that connection.

- Any open connection MUST be kept alive by the client until the Snap explicitly closes it or the connection is lost, even when the Snap is terminated.

- When the Snap requests to close a WebSocket connection via `snap_closeWebSocket`, the client MUST close the connection and remove it from its list of open connections.

- Any errors that occur when the WebSocket connection is open (e.g., connection failure, message send failure) MUST be handled by the client and reported to the Snap via `onWebSocketEvent`.

### Snap Manifest

This SIP doesn't introduce any new permissions, but rather extends `endowment:network-access` with new capabilities.

### RPC Methods

#### `snap_openWebSocket`

> This method allows a Snap to open a WebSocket connection to an external service. The method takes a URL as a parameter and returns a unique identifier for the connection.

Reviewer 3: Can the WebSocket be used immediately after calling `snap_openWebSocket`, or do we need to wait for the `"open"` event?

```typescript

type OpenWebSocketParams = {
  url: string;
  protocols?: string[];
};
```
The RPC method takes one parameter:

- `url` - The URL of the WebSocket service to connect to. 
  - The URL MUST be a valid WebSocket URL, starting with `wss://`. URLs starting with `ws://` are not allowed.
  - Only one WebSocket connection can be opened per URL at a time. If a Snap tries to open a new connection to the same URL while an existing connection is still open, the client MUST throw an error.

> - `protocols` - An optional array of subprotocols to use for the WebSocket connection.

Reviewer 3: What is the usage for this field?

Author: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#protocols

Reviewer 3: Just a note: if, for some reason, the Snap needs to use two sub-protocols, it will need to open two connections:

> Note however that only one sub-protocol can be selected per connection.

  message: string;
};
export type WebSocketBinaryMessage = {
  type: "binary";
  message: number[];
};

export type WebSocketMessageData =
  | WebSocketTextMessage
  | WebSocketBinaryMessage;

export type WebSocketMessage = {
  type: "message";
  id: string;
  origin: string;
  data: WebSocketMessageData;
};
```

`type` - The type of the WebSocket event, which is `"message"` for this event type.

`id` - The unique identifier of the WebSocket connection associated with the event.

`origin` - The origin of the WebSocket event.

`data` - The data received in the event. This can be either a text message (as a string) or binary data (as an array of numbers). The `type` property indicates whether the data is text or binary.

##### Connection opened event

This event is triggered when a WebSocket connection is successfully opened. It provides the unique identifier of the connection and the origin of the event.

```typescript
export type WebSocketOpenEvent = {
  type: "open";
  id: string;
  origin: string;
};
```
`type` - The type of the WebSocket event, which is `"open"` for this event type.

`id` - The unique identifier of the WebSocket connection associated with the event.

> `origin` - The origin of the WebSocket connection.

Author: The WebSocket origin for easy access to what origin you have open, similar to https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/origin

> `origin` - The origin of the WebSocket connection.

Reviewer 3: It's clear in the case of a `MessageEvent`:

> ... origin of the message emitter.

But in the case of a `WebSocketOpenEvent` event, the `origin` is the Snap. Is it correct?

##### Connection closed event

This event is triggered when a WebSocket connection is closed. It provides the unique identifier of the connection and the origin of the event.

```typescript
export type WebSocketCloseEvent = {
  type: "close";
  id: string;
  origin: string;
  code: number;
  reason: string;
  wasClean: boolean;
};
```

`type` - The type of the WebSocket event, which is `"close"` for this event type.

`id` - The unique identifier of the WebSocket connection associated with the event.

`origin` - The origin of the WebSocket connection.

`code` - The numeric code indicating the reason for the closure of the WebSocket connection. This is a standard WebSocket close code.

`reason` - A string providing a human-readable explanation of why the WebSocket connection was closed.

`wasClean` - A boolean indicating whether the connection was closed cleanly (i.e., without any errors).

##### Error event

This event is triggered when an error occurs with a WebSocket connection. It provides the unique identifier of the connection, and the origin of the event.

```typescript
export type WebSocketErrorEvent = {
  type: "error";
  id: string;

>   origin: string;

Reviewer 2: Yes but a snap might keep in state the id <-> origin correlation so it might be easier for it to check the origin

};
```
`type` - The type of the WebSocket event, which is `"open"` for this event type.

`id` - The unique identifier of the WebSocket connection associated with the event.

`origin` - The origin of the WebSocket connection.

##### Connection closed event

This event is triggered when a WebSocket connection is closed. It provides the unique identifier of the connection and the origin of the event.

```typescript
export type WebSocketCloseEvent = {
  type: "close";
  id: string;
  origin: string;
  code: number;
  reason: string;
};
```

`type` - The type of the WebSocket event, which is `"close"` for this event type.

`id` - The unique identifier of the WebSocket connection associated with the event.

`origin` - The origin of the WebSocket connection.

`code` - The numeric code indicating the reason for the closure of the WebSocket connection. This is a standard WebSocket close code.

`reason` - A string providing a human-readable explanation of why the WebSocket connection was closed.

##### Error event

This event is triggered when an error occurs with a WebSocket connection. It provides the unique identifier of the connection, and the origin of the event.

```typescript

> export type WebSocketErrorEvent = {

Reviewer: This should probably include the error? Or does the WebSocket API not give an actual error message?

Reviewer 2: The `error` event don't give any error message, the event type for errors is a generic `Event`. You get a code and a reason in the `close` event, that's why I just added these properties to the close event

Reviewer 2: https://stackoverflow.com/a/18804298

Author: Do we need to emit this event at all then?

Author: It might not be useful if it provides no information at all 😅

Author: Hmm, I was assuming an error would always close, but maybe not 🤷‍♂️

Reviewer 2: This would allow us to be close to the og. WS

sip: 20

> title: External data entry point

Author: WebSockets

> title: External data entry point

Author: ```suggestion
title: WebSockets
```

- `url` - The URL of the WebSocket service to connect to. 
  - The URL MUST be a valid WebSocket URL, starting with `wss://`. URLs starting with `ws://` are not allowed.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_openWebSocket",
  params: {
    url: "wss://example.com/websocket",
  },
});

```

#### `snap_closeWebSocket`
This method allows a Snap to close an existing WebSocket connection. The method takes the unique identifier of the connection as a parameter.

```typescript
type CloseWebSocketParams = {
  id: string;
};
```
The RPC method takes one parameter:
- `id` - The unique identifier of the WebSocket connection to close. This identifier is returned by the `snap_openWebSocket` method.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_closeWebSocket",
  params: {
    id: "unique-connection-id",
  },
});
```
#### `snap_sendMessage`
This method allows a Snap to send a message over an existing WebSocket connection. The method takes the unique identifier of the connection and the message to send as parameters.

```typescript
type SendWebSocketMessageParams = {
  id: string;

>   message: string | Uint8Array;

Reviewer 2: Yeah let's start here and we can always update later

};
```

The RPC method takes two parameters:
- `id` - The unique identifier of the WebSocket connection to send the message over. This identifier is returned by the `snap_openWebSocket` method.
- `message` - The message to send over the WebSocket connection. It can be either a string or a `Uint8Array`.

An example of usage is given below.

```typescript
snap.request({
  method: "snap_sendMessage",
  params: {
    id: "unique-connection-id",
    message: "Hello, WebSocket!",
  },
});
```
#### `snap_getWebSockets`
This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier and URL of the connection.

- `id` - The unique identifier of the WebSocket connection.
- `url` - The URL of the WebSocket connection.

```typescript
type WebSocketConnection = {
  id: string;

>   url: string;

Author: We should choose either URL or origin and use that everywhere

Reviewer 2: I agree that for open connections `origin` is the way to go but to open a connection like in `snap_openWebSocket` we should use `url` as `origin` doesn't make sense

      break;
  }
};
```
the type for an `onWebSocketEvent` handler function's arguments is:

```typescript
export type WebSocketMessageEvent = {
  type: "message";
  id: string;
  origin: string;
  dataType: "text" | "binary";
  data: string | Uint8Array;
};

export type WebSocketOpenEvent = {
  type: "open";
  id: string;
  origin: string;
};

export type WebSocketCloseEvent = {
  type: "close";
  id: string;
  origin: string;
};

export type WebSocketErrorEvent = {
  type: "error";
  id: string;
  origin: string;
};

export type WebSocketEvent =
  | WebSocketDataEvent
  | WebSocketOpenEvent
  | WebSocketCloseEvent
  | WebSocketErrorEvent;
```

```typescript
type OnWebSocketEventArgs = {
  event: WebSocketEvent;
};
```
`type` - The type of the WebSocket event, which can be one of the following:
  - `message` - Indicates that a message has been received.
  - `open` - Indicates that a WebSocket connection has been opened.
  - `close` - Indicates that a WebSocket connection has been closed.
  - `error` - Indicates that an error has occurred with the WebSocket connection.

`id` - The unique identifier of the WebSocket connection associated with the event.

> `origin` - The origin of the Snap that is handling the WebSocket event.

Author: ```suggestion
`origin` - The origin of the WebSocket event.
```
